### PR TITLE
Update usage of FileInput/OutputStream

### LINKED ISSUE #1835: FileInputStream / FileOutputStream Considered Harmful
https://groups.google.com/d/msg/liftweb/wbUbxzQsyJs/O8zvu2lDAQAJ

Replace usage of ```new FileInputStream``` and ```new FileOutputStream``` with newer java.nio equivalents

### PR DESCRIPTION
**Pull request by eltimn**
Fixes #1835 

I haven't been able to directly test this yet, but I started to upgrade our app to use Scala 2.12 and will as soon as that's done.

### CODE REVIEW

File: core/util/src/main/scala/net/liftweb/util/IoHelpers.scala [hunk 1]
```diff
@@ -85,7 +86,8 @@ trait IoHelpers {
   /**
    * Read an entire file into an Array[Byte]
    */
-  def readWholeFile(file: File): Array[Byte] = readWholeStream(new FileInputStream(file))
+  def readWholeFile(file: File): Array[Byte] = readWholeStream(Files.newInputStream(file.toPath))
+  def readWholeFile(path: Path): Array[Byte] = readWholeStream(Files.newInputStream(path))
```
Comment (farmdawgnation) [inline core/util/src/main/scala/net/liftweb/util/IoHelpers.scala]: Will the scaladoc appear for both methods if they're grouped like this or should we copy it to be above the second instance, too?

Comment (eltimn) [inline core/util/src/main/scala/net/liftweb/util/IoHelpers.scala]: I'm not sure. I'll try it locally and see what it looks like.

Comment (eltimn) [inline core/util/src/main/scala/net/liftweb/util/IoHelpers.scala]: It did *not* appear, so I copied it.

**Comment by farmdawgnation** [comment]
@eltimn Are we waiting on a first-hand testing report from you or is this good to merge?

**Comment by eltimn** [comment]
I did get our app upgraded to scala 2.12, but I'd need to publish a few lift modules compiled against 3.1 before I could test this.

I did make these same changes to our app without issues, so I'm good with merging it.

**Comment by eltimn** [comment]
I added a spec to test ```readWholeFile```, which is what I mainly wanted to make sure still worked properly.

**Comment by farmdawgnation** [comment]
Sweet. I say we ship this, then.

File: web/webkit/src/main/scala/net/liftweb/http/Req.scala [hunk 2]
```diff
@@ -284,14 +284,14 @@ FileParamHolder(name, mimeType, fileName)
  * @param localFile The local copy of the uploaded file
  */
 class OnDiskFileParamHolder(override val name: String, override val mimeType: String,
-                            override val fileName: String, val localFile: File) extends
+                            override val fileName: String, val localFile: Path) extends
```
Comment (Shadowfiend) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: Whelp, I missed this guy. This will break all consumers of `OnDiskFileParamHolder`. Is there a way to get from `Path` to `File`? If not, I suggest we revert the changes to `OnDiskFileParamHolder`. If yes, I suggest we rename this parameter `localPath` and provide a `localFile` accessor that returns a `File` as existing code will expect. We can mark that one deprecated if we really want to.

Comment (eltimn) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: There's an apply function with File, so it shouldn't break anything. Also, there's a toPath function on File, and a toFile function on Path that can convert either way.

Comment (Shadowfiend) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: See https://github.com/lift/framework/pull/1852#issuecomment-302056802 . Folks are using `.localFile` directly, and passing it to things that expect a `File`. Moreover, `localFile` is a poor name for something that isn't a `File`, but rather a `Path`, IMO.

I'll work up a PR to make this API-compatible, should only be a couple of lines.

Comment (eltimn) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: I just realized that's not a case class, though. Do I need to change that to a constructor?

Comment (eltimn) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: I see what I did wrong now. Thanks for fixing it.

Comment (Shadowfiend) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: I'll probably add a constructor overload as well, good point. I also missed that it wasn't a case class hah. I doubt anyone is using the constructor directly, but better safe than sorry.

Comment (Shadowfiend) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: Heh, so the reason `FileInputStream` and `FileOutputStream` are “considered harmful” in the Cloudbees post is because they use finalizers. But `OnDiskFileParamHolder` also has a finalizer. I think this is probably still a win, but worth mentioning.

Comment (eltimn) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: The finalize function is on OnDiskFileParamHolder, not on the FileStream class, though. I would assume that's a bit different, although I guess I don't know exactly how having finalizers affect GC.

Comment (Shadowfiend) [inline web/webkit/src/main/scala/net/liftweb/http/Req.scala]: Yeah not sure either. Will keep it in mind in case anything around this comes up in the future.
